### ISSUE TITLE
[BUG] Event Hubs service rejecting SAS token for unknown reason

### ISSUE THREAD

**Issue by js8080** [Azure/azure-sdk-for-net#26136]
### Library name and version

Azure.Messaging.EventHub 5.6.2

### Describe the bug

I am using the `Azure.Messaging.EventHubs.Producer.EventHubProducerClient` from my .NET 6 Web project to publish events to an Azure Event Hub. This was working fine when I used a connection string to authenticate the `EventHubProducerClient` but I want to switch to using SharedAccessSignature Tokens instead.

Unfortunately, when I try to use SAS tokens, I get this error:
```
System.UnauthorizedAccessException: Put token failed. status-code: 401, status-description: Malformed authorization token. 
```

### Expected behavior

Correctly authenticate when using SAS tokens.

### Actual behavior

An error is returned:
```
System.UnauthorizedAccessException: Put token failed. status-code: 401, status-description: Malformed authorization token. 
```

### Reproduction Steps

I am registering my `EventHubProdocuerClient` as a Scoped lifetime service in my service provider. I am reading the SAS token from a file (which will actually be mounted to my pod as a Kubernetes secret but that detail isn't important here):
```csharp
    services.AddScoped<Azure.Messaging.EventHubs.Producer.EventHubProducerClient>((sp) =>
    {
        string sasToken = fileReader.ReadAllText(sharedAccessSignatureFilePath).Result;
        
        return new Azure.Messaging.EventHubs.Producer.EventHubProducerClient(fullyQualifiedNamespace,
            eventHubName,
            new AzureSasCredential(sasToken)
    });
```

`fullyQualifiedNamespace` is like this: 
```
my-eventhub-001.servicebus.windows.net
```

First I created a Shared Access Policy in my Event Hub in the Azure Portal called "EventProducer" and I gave it "Send" rights. If I click on that, it shows me a connection string and keys. I pulled some of the info like the resourceUri, keyName, and key from there.

I generated my SAS token using the C# code example as shown in [Authenticate access to Event Hubs resources using shared access signatures (SAS) -> Generate a Shared Access Signature token -> C#](https://docs.microsoft.com/en-us/azure/event-hubs/authenticate-shared-access-signature#c)

My SAS token looks like this:
```
SharedAccessSignature sr=sb%3a%2f%2fmy-eventhub-001.servicebus.windows.net%2fmyhub1&sig=9q3N{...omitted...}KJy78%3d&se=1640997904&skn=EventProducer
```

Now I try to create a message batch with my `EventHubProducerClient:
```csharp
    using EventDataBatch eventBatch = await _producerClient.CreateBatchAsync();
```

It fails with the following exception:
```
{System.UnauthorizedAccessException: Put token failed. status-code: 401, status-description: Malformed authorization token. TrackingId:775b1ad2-1250-4523-b840-4c7781cb6522_G7, SystemTracker:my-eventhub-001.servicebus.windows.net:myhub1, Timestamp:2021-12-31T23:12:12.
 ---> Microsoft.Azure.Amqp.AmqpException: Put token failed. status-code: 401, status-description: Malformed authorization token. TrackingId:775b1ad2-1250-4523-b840-4c7781cb6522_G7, SystemTracker:my-eventhub-001.servicebus.windows.net:myhub1, Timestamp:2021-12-31T23:12:12.
   at Microsoft.Azure.Amqp.ExceptionDispatcher.Throw(Exception exception)
   at Microsoft.Azure.Amqp.AsyncResult.End[TAsyncResult](IAsyncResult result)
   at Microsoft.Azure.Amqp.AsyncResult`1.End(IAsyncResult asyncResult)
   at Microsoft.Azure.Amqp.AmqpCbsLink.EndSendToken(IAsyncResult result)
   at Microsoft.Azure.Amqp.AmqpCbsLink.<>c__DisplayClass4_0.<SendTokenAsync>b__1(IAsyncResult a)
   at System.Threading.Tasks.TaskFactory`1.FromAsyncCoreLogic(IAsyncResult iar, Func`2 endFunction, Action`1 endAction, Task`1 promise, Boolean requiresSynchronization)
--- End of stack trace from previous location ---
   at Azure.Messaging.EventHubs.Amqp.AmqpConnectionScope.RequestAuthorizationUsingCbsAsync(AmqpConnection connection, CbsTokenProvider tokenProvider, Uri endpoint, String audience, String resource, String[] requiredClaims, TimeSpan timeout)
   at Azure.Messaging.EventHubs.Amqp.AmqpConnectionScope.CreateSendingLinkAsync(AmqpConnection connection, Uri endpoint, TransportProducerFeatures features, PartitionPublishingOptionsInternal options, TimeSpan timeout, String linkIdentifier, CancellationToken cancellationToken)
   at Azure.Messaging.EventHubs.Amqp.AmqpConnectionScope.OpenProducerLinkAsync(String partitionId, TransportProducerFeatures features, PartitionPublishingOptionsInternal options, TimeSpan timeout, String linkIdentifier, CancellationToken cancellationToken)
   at Azure.Messaging.EventHubs.Amqp.AmqpProducer.CreateLinkAndEnsureProducerStateAsync(String partitionId, String producerIdentifier, PartitionPublishingOptionsInternal partitionOptions, TimeSpan timeout, CancellationToken cancellationToken)
   at Azure.Messaging.EventHubs.Amqp.AmqpProducer.CreateLinkAndEnsureProducerStateAsync(String partitionId, String producerIdentifier, PartitionPublishingOptionsInternal partitionOptions, TimeSpan timeout, CancellationToken cancellationToken)
   at Microsoft.Azure.Amqp.FaultTolerantAmqpObject`1.OnCreateAsync(TimeSpan timeout)
   at Microsoft.Azure.Amqp.Singleton`1.GetOrCreateAsync(TimeSpan timeout)
   at Microsoft.Azure.Amqp.Singleton`1.GetOrCreateAsync(TimeSpan timeout)
   at Azure.Messaging.EventHubs.Amqp.AmqpProducer.CreateBatchAsync(CreateBatchOptions options, CancellationToken cancellationToken)
   --- End of inner exception stack trace ---
   at Azure.Messaging.EventHubs.Amqp.AmqpProducer.CreateBatchAsync(CreateBatchOptions options, CancellationToken cancellationToken)
   at Azure.Messaging.EventHubs.Producer.EventHubProducerClient.CreateBatchAsync(CreateBatchOptions options, CancellationToken cancellationToken)
   at Azure.Messaging.EventHubs.Producer.EventHubProducerClient.CreateBatchAsync(CancellationToken cancellationToken)
   at Epic.ApiGateway.Diagnostics.TestPerformer.ResultsReporting.AzureEventHubResultReporter.Report(DestinationTestResult testResult) in /home/justin/source/repos/src/MyProject/AzureEventHubEventProducer.cs:line 34}
```

I have tried changing the connection options in my producer client to use WebSockets like this but that doesn't help:
```csharp
    services.AddScoped<Azure.Messaging.EventHubs.Producer.EventHubProducerClient>((sp) =>
    {
        string sasToken = fileReader.ReadAllText(sharedAccessSignatureFilePath).Result;
        
        return new Azure.Messaging.EventHubs.Producer.EventHubProducerClient(fullyQualifiedNamespace,
            eventHubName,
            new AzureSasCredential(sasToken),
            new EventHubProducerClientOptions()
            {
                ConnectionOptions = new EventHubConnectionOptions()
                {
                    TransportType = EventHubsTransportType.AmqpWebSockets
                }
            });
    });
```

I also tried taking the SAS token and manually sending a request to the Event Hub using HTTPS and that seemed to work just fine:

```
curl -v -H 'Authorization: SharedAccessSignature sr=https%3a%2f%2fmy-eventhub-001.servicebus.windows.net%2fmyhub1&sig=KlJ4U1{...omitted...}g%3d&se=1640988523&skn=EventProducer' --data 'hello world!' https://my-eventhub-001.servicebus.windows.net/myhub1/messages?timeout=60\&api-version=2014-01
```

This returns a 201:
```
< HTTP/1.1 201 Created
```

It seems to me there is a bug in `Azure.Messaging.EventHubs.Producer.EventHubProducerClient` which makes it not work with SAS tokens... Unless I am using it incorrectly in which case the documentation should be updated to clarify the proper use. On that note, I also created an issue on the documentation page about some of the code there: https://github.com/MicrosoftDocs/azure-docs/issues/85870

### Environment

dotnet --info:
```
.NET SDK (reflecting any global.json):
 Version:   6.0.100
 Commit:    9e8b04bbff

Runtime Environment:
 OS Name:     ubuntu
 OS Version:  20.04
 OS Platform: Linux
 RID:         ubuntu.20.04-x64
 Base Path:   /usr/share/dotnet/sdk/6.0.100/

Host (useful for support):
  Version: 6.0.0
  Commit:  4822e3c3aa

.NET SDKs installed:
  3.1.416 [/usr/share/dotnet/sdk]
  5.0.403 [/usr/share/dotnet/sdk]
  6.0.100 [/usr/share/dotnet/sdk]

.NET runtimes installed:
  Microsoft.AspNetCore.App 3.1.22 [/usr/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 5.0.12 [/usr/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 6.0.0 [/usr/share/dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 3.1.22 [/usr/share/dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 5.0.12 [/usr/share/dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 6.0.0 [/usr/share/dotnet/shared/Microsoft.NETCore.App]

To install additional .NET runtimes or SDKs:
  https://aka.ms/dotnet-download
```

IDE is VS Code
```
1.63.2
899d46d82c4c95423fb7e10e68eba52050e30ba3
x64
```

**Comment by jsquire** [comment]
Hi @js8080.  Thank you for reaching out and we regret that you're experiencing difficulties.  The Event Hubs client library does support SAS tokens and the functionality is validated as part of our nightly test runs.  I double-checked test results and am not seeing failures.  I also confirmed that I'm able to publish using a token generated with the `createToken` token function in the [documentation](https://docs.microsoft.com/azure/event-hubs/authenticate-shared-access-signature) that you referenced without issue.  

Based on the message and stack trace in the error that you're seeing, Event Hubs service is rejecting your token for some reason.  The format of the token is correct; it passed client validation and was able to be parsed into its components.  It may be that the key itself is malformed, it may be that the service error text is misleading and the token is valid but not authorized for `myhub1`, or it may be neither of those.   

At this point, we'll need the Event Hubs service team to assist so that we can understand the service behavior and exactly why the token was rejected.   I've tagged this and routed it to their attention.  Please forgive delays in response, as many folks are out for the holidays through this week.

**Comment by jsquire** [comment]
//cc: @serkantkaraca, @JamesBirdsall, @sjkwak

**Comment by js8080** [comment]
I replaced the real names of my event hub for privacy reasons but are these representative values for the parameters to `public EventHubProducerClient(string fullyQualifiedNamespace, string eventHubName, AzureSasCredential credential, EventHubProducerClientOptions clientOptions = null)`?

- `fullyQualifiedNamespace` - "my-eventhub-001.servicebus.windows.net"
- `eventHubName` - 'myhub1`

And then for `private static string createToken(string resourceUri, string keyName, string key)` ([in the documentation](https://docs.microsoft.com/en-us/azure/event-hubs/authenticate-shared-access-signature#c)):

- `resourceUri` - 'sb://my-eventhub-001.servicebus.windows.net'
- `keyName` - 'EventPublisher' - Here I put the name of my Shared access policy from the Azure portal -> Go to my Event Hubs Namespace -> Event Hubs -> select my hub "myhub1" -> Shared Access Policies
- `key` - Here I put the primary key from the Azure portal -> Go to my Event Hubs Namespace -> Event Hubs -> select my hub "myhub1" -> Shared Access Policies -> go to the policy I created called 'EventProducer' and then on the right I click the copy button under 'Primary key'.

Perhaps I messed something up with the values I'm using for these params.

Another possibility is the expiration date I'm using. Is there a time limit on the expiration? I slightly modified the code in [in the documentation](https://docs.microsoft.com/en-us/azure/event-hubs/authenticate-shared-access-signature#c) to take a parameter for expiration time: and in my case I have been passing in a TimeSpan of 6000 seconds (100 minutes).

```csharp
        // https://docs.microsoft.com/en-us/azure/event-hubs/authenticate-shared-access-signature#c
        private static string CreateTokenManually(string resourceUri, string keyName, string key, TimeSpan expiresIn)
        {
            TimeSpan sinceEpoch = DateTime.UtcNow - new DateTime(1970, 1, 1);
            var expiry = Convert.ToString((int)sinceEpoch.TotalSeconds + expiresIn.TotalSeconds);
            string stringToSign = HttpUtility.UrlEncode(resourceUri) + "\n" + expiry;
            using (var hmac = new HMACSHA256(Encoding.UTF8.GetBytes(key)))
            {
                var signature = Convert.ToBase64String(hmac.ComputeHash(Encoding.UTF8.GetBytes(stringToSign)));
                var sasToken = String.Format(CultureInfo.InvariantCulture, "SharedAccessSignature sr={0}&sig={1}&se={2}&skn={3}", HttpUtility.UrlEncode(resourceUri), HttpUtility.UrlEncode(signature), expiry, keyName);
                return sasToken;
            }
        }
```
